src/index.ts

```diff
@@ -1,53 +1,266 @@
+          path: args['cwd'],
+          type: 'file',
+          ext: 'ts'
+        });
+        const widget: IDocumentWidget<FileEditor> | undefined =
+          await app.commands.execute('docmanager:open', {
+            path: model.path,
+            factory: 'Editor'
+          });
+        if (widget) {
+          widget.content.ready.then(() => {
+            widget.content.model.value.text = PLUGIN_TEMPLATE;
+          });
+        }
+        return widget;
+      }
+    });
+
+    app.restored.then(async () => {
+      const settings = this.settings;
+      this._updateSettings(requirejs, settings);
+      // add to the launcher
+      if (launcher && (settings.composite.showIconInLauncher as boolean)) {
+        launcher.add({
+          command: CommandIDs.createNewFile,
+          category: 'Other',
+          rank: 1
+        });
+      }
+
+      const urls = settings.composite.urls as string[];
+      for (const u of urls) {
+        await this._getModule(u);
+      }
+      const plugins = settings.composite.plugins as string[];
+      for (const t of plugins) {
+        await this._loadPlugin(t, null);
+      }
+
+      settings.changed.connect(updatedSettings => {
+        this._updateSettings(requirejs, updatedSettings);
+      });
+    });
   }
-  // We allow one level of indirection (return a function instead of a plugin)
-  if (typeof plugin === 'function') {
-    plugin = plugin();
+
+  private _updateSettings(
+    requirejs: IRequireJS,
+    settings: ISettingRegistry.ISettings
+  ) {
+    const baseURL = settings.composite.requirejsCDN as string;
+    requirejs.require.config({
+      baseUrl: baseURL
+    });
   }
 
-  // Finally, we allow returning a promise (or an async function above).
-  plugin = await Promise.resolve(plugin);
+  private async _loadPlugin(code: string, path: string | null) {
+    const tokenMap = new Map(
+      Array.from((this.app as any)._serviceMap.keys()).map((t: any) => [
+        t.name,
+        t
+      ])
+    );
+    // Widget registry does not follow convention of importName:tokenName
+    tokenMap.set(
+      '@jupyter-widgets/base:IJupyterWidgetRegistry',
+      tokenMap.get('jupyter.extensions.jupyterWidgetRegistry')
+    );
+    const importResolver = new ImportResolver({
+      modules: modules as unknown as Record<string, IModule>,
+      tokenMap: tokenMap,
+      requirejs: this.requirejs,
+      settings: this.settings,
+      serviceManager: this.app.serviceManager,
+      basePath: path
+    });
+
+    const pluginLoader = new PluginLoader({
+      transpiler: new PluginTranspiler({
+        compilerOptions: {
+          target: ts.ScriptTarget.ES2017,
+          jsx: ts.JsxEmit.React
+        }
+      }),
+      importFunction: importResolver.resolve.bind(importResolver),
+      tokenMap: tokenMap,
+      serviceManager: this.app.serviceManager,
+      requirejs: this.requirejs
+    });
+    importResolver.dynamicLoader = pluginLoader.loadFile.bind(pluginLoader);
```

Comment:
Maybe rename `importFunction` to `importResolver` for coherence.

Reply:
I differentiate them to distinguish between the `.resolve` method and the `ImportResolver` class.